[sc-59665] v0.37 changelog (#5928)
PennyLaneAI/pennylane: doc/releases/changelog-0.37.0.md

  [(#5744)](https://github.com/PennyLaneAI/pennylane/pull/5744)
  [(#5786)](https://github.com/PennyLaneAI/pennylane/pull/5786)
  [(#5795)](https://github.com/PennyLaneAI/pennylane/pull/5795)

  Either method can be selected when instantiating the `default.tensor` device by setting the
  `method` keyword argument to `"tn"` (tensor network) or `"mps"` (matrix product state). This
  device can simulate a large number of qubits. Take this example that
  simulates 1000 qubits with a shallow circuit in a few seconds!

  ```python
  import pennylane as qml

  num_qubits = 1000
  dev = qml.device("default.tensor", method="mps")

  @qml.qnode(dev)
  def circuit():
      qml.Hadamard(0)
      for i in range(num_qubits - 1):
          qml.CNOT([i, i + 1])
      return qml.expval(qml.Z(num_qubits - 1))
  ```

  ```pycon
  >>> circuit()
  tensor(0., requires_grad=True)
  ```

  For matrix product state simulations (`method="mps"`), we can make the exeuction approximate by setting `max_bond_dim` (see the
  [device's documentation](https://docs.pennylane.ai/en/stable/code/api/pennylane.devices.default_tensor.DefaultTensor.html)
  for more details).

>   The maximum bond dimension lets us control the degree of the approximation and has implications for the speed of the simulation, as shown in the

Author: Just rewording a bit because we don't show any timing results

```suggestion
  The maximum bond dimension has implications for the speed of the simulation and lets us control the degree of the approximation, as shown in the
```

  analyzing quantum circuits during execution.
  [(#5680)](https://github.com/PennyLaneAI/pennylane/pull/5680)
  [(#5749)](https://github.com/PennyLaneAI/pennylane/pull/5749)
  [(#5789)](https://github.com/PennyLaneAI/pennylane/pull/5789)
  
  This includes monitoring the circuit via measurements using `qml.debug_state()`, `qml.debug_probs()`, 
  `qml.debug_expval()`, and `qml.debug_tape()`, stepping through the operations in a quantum circuit, and interactively
  adding operations during execution.

  Including `qml.breakpoint()` in a circuit will cause the simulation to pause during execution and bring up the interactive console.
  For example, consider the following code in a Python file called `script.py`:

  ```python
  @qml.qnode(qml.device('default.qubit', wires=(0,1,2)))
  def circuit(x):
      qml.Hadamard(wires=0)
      qml.CNOT(wires=(0,2))
      qml.breakpoint()

      qml.RX(x, wires=1)
      qml.RY(x, wires=2)
      qml.breakpoint()

      return qml.sample()

  circuit(1.2345)
  ```

  Upon executing `script.py`, the simulation pauses at the first breakpoint:

> 

Author: ```suggestion
```

  (pldb): tape.operations
  [Hadamard(wires=[0]), CNOT(wires=[0, 2])]
  ```

  `qml.debug_state()` is equivalent to `qml.state()` and gives the current state:

  ```pycon
  (pldb): print(qml.debug_state())
  [0.70710678+0.j 0.        +0.j 0.        +0.j 0.        +0.j
    1.        +0.j 0.70710678+0.j 0.        +0.j 0.        +0.j]
  ```

  Other debugger functions like `qml.debug_probs()` and `qml.debug_expval()` also function like their simulation counterparts (`qml.probs` and `qml.expval`, respectively) and are described in more detail in the [debugger documentation](https://docs.pennylane.ai/en/stable/code/qml_debugging.html)
  
  Additionally, standard debugging commands are available to navigate through code, including `list`, `longlist`, `next`, `continue`, and `quit` as described in [the debugging documentation](https://docs.pennylane.ai/en/stable/code/qml_debugging.html#controlling-code-execution-in-the-debugging-context).
  
  Finally, to modify a circuit mid-run, simply call the desired PennyLane operations:

  ```pycon
  (pldb) qml.CNOT(wires=(0,2))
  CNOT(wires=[0, 2])
  (pldb): print(qml.debug_tape().draw(wire_order=[0,1,2]))
  0: ──H─╭●─╭●─┤  
  2: ────╰X─╰X─┤  

>   ```
> 

Author: We have a demo coming soon, right?

```suggestion
  ```

Stay tuned for an in-depth demonstration on using this feature in real-world examples!
```

>   ```
> 

Author: We have a demo coming soon, right?

```suggestion

Stay tuned for an in-depth demonstration on using this feature with real-world examples!
```

> 

Author: We have a demo coming soon, right?

```suggestion

Stay tuned for an in-depth demonstration on using this feature with real-world examples!
```

  0: ──H─┤  <X>

  0: ──H─┤  <Y>
  ```

  The
  [`qml.workflow.get_transform_program()` function](https://docs.pennylane.ai/en/latest/code/api/pennylane.workflow.get_transform_program.html)
  can be used to see the full transform program.

  ```pycon
  >>> qml.workflow.get_transform_program(f)
  TransformProgram(merge_rotations, cancel_inverses, split_non_commuting, validate_device_wires, mid_circuit_measurements, decompose, validate_measurements, validate_observables, no_sampling)
  ```

  Note that additional transforms can be added automatically from device preprocessing or gradient
  calculations. Rather than providing an integer value to `level`, it is possible to target
  the `"user"`, `"gradient"` or `"device"` stages:

  ```python
  n_wires = 3
  x = np.random.random((2, n_wires))

  @qml.qnode(qml.device("default.qubit"))
  def f():
      qml.BasicEntanglerLayers(x, range(n_wires))
      return qml.expval(qml.X(0))
  ```

  ```pycon
  >>> print(qml.draw(f, level="device")())
  0: ──RX(0.28)─╭●────╭X──RX(0.70)─╭●────╭X─┤  <X>
  1: ──RX(0.52)─╰X─╭●─│───RX(0.65)─╰X─╭●─│──┤     
  2: ──RX(0.00)────╰X─╰●──RX(0.03)────╰X─╰●─┤     

>   ```

Author: Nice!

  [tensor network](https://en.wikipedia.org/wiki/Tensor_network) and
  [matrix product state](https://en.wikipedia.org/wiki/Matrix_product_state) simulations
  of quantum circuits using the
  [quimb backend](https://quimb.readthedocs.io/en/latest/).
  [(#5699)](https://github.com/PennyLaneAI/pennylane/pull/5699)
  [(#5744)](https://github.com/PennyLaneAI/pennylane/pull/5744)
  [(#5786)](https://github.com/PennyLaneAI/pennylane/pull/5786)
  [(#5795)](https://github.com/PennyLaneAI/pennylane/pull/5795)

  Either method can be selected when instantiating the `default.tensor` device by setting the
  `method` keyword argument to `"tn"` (tensor network) or `"mps"` (matrix product state). This
  device can simulate a large number of qubits. Take this example that
  simulates 1000 qubits with a shallow circuit in a few seconds!

  ```python
  import pennylane as qml

  num_qubits = 1000
  dev = qml.device("default.tensor", method="mps")

  @qml.qnode(dev)
  def circuit():
      qml.Hadamard(0)
      for i in range(num_qubits - 1):
          qml.CNOT([i, i + 1])
      return qml.expval(qml.Z(num_qubits - 1))
  ```

  ```pycon
  >>> circuit()
  tensor(0., requires_grad=True)
  ```

>   For matrix product state simulations (`method="mps"`), we can make the exeuction approximate by setting `max_bond_dim` (see the

Reviewer: ```suggestion
  For matrix product state simulations (`method="mps"`), we can make the execution be approximate by setting `max_bond_dim` (see the
```

  ```

  The `default.tensor` device is instantiated with a `max_bond_dim` value:

  ```python
  dev_dq = qml.device("default.qubit")
  value_dq = qml.QNode(f, dev_dq)()

  dev_mps = qml.device("default.tensor", max_bond_dim=5)
  value_mps = qml.QNode(f, dev_mps)()
  ```

  With this bond dimension, the expectation values calculated for `default.qubit` and
  `default.tensor` are different:

  ```pycon
  >>> np.abs(value_dq - value_mps)
  tensor(0.0253213, requires_grad=True)
  ```

  Learn more about `default.tensor` and how to configure it by 
  [visiting the how-to guide](https://pennylane.ai/qml/demos/tutorial_How_to_simulate_quantum_circuits_with_tensor_networks/).

<h4>Add noise models to your quantum circuits 📺</h4>

* Support for building noise models and applying them to a quantum circuit has been added
via the `NoiseModel` class and an `add_noise` transform.
  [(#5674)](https://github.com/PennyLaneAI/pennylane/pull/5674)
  [(#5684)](https://github.com/PennyLaneAI/pennylane/pull/5684)
  [(#5718)](https://github.com/PennyLaneAI/pennylane/pull/5718)

  Under the hood, PennyLane's approach to noise models is insertion-based, meaning that noise is included

>   by *inserting* additional operators (gates or channels) that describe the noise into the quantum circuit itself. 

Reviewer: ```suggestion
  by *inserting* additional operators (gates or channels) that describe the noise into the quantum circuit.
```

  Creating a `NoiseModel` boils down to defining Boolean conditions under which specific noisy operations 
  are inserted. There are several ways to specify conditions for adding noisy operations: 

  * `qml.noise.op_eq(op)`: if the operator `op` is encountered in the circuit, add noise.
  * `qml.noise.op_in(ops)`: if any operators in `ops` are encountered in the circuit, add noise.
  * `qml.noise.wires_eq(wires)`: if an operator is applied to `wires`, add noise.
  * `qml.noise.wires_in(wires)`: if an operator is applied to any wire in `wires`, add noise.
  * custom noise conditions: custom conditions can be defined as functions decorated with `qml.BooleanFn` that return a Boolean value. For example, the following function will insert noise if a `qml.RY` operator is encountered with an angle of rotation that is less than `0.5`:

    ```python
    @qml.BooleanFn
    def c0(op):
        return isinstance(op, qml.RY) and op.parameters[0] < 0.5
    ```

>   `op_eq`, `op_in`, `wires_eq`, and `wires_in` can also be combined together with `&`, `and`, `|`, etc.

Reviewer: ```suggestion
  Conditions can also be combined together with `&`, `and`, `|`, etc.
```

  
  Additionally, standard debugging commands are available to navigate through code, including `list`, `longlist`, `next`, `continue`, and `quit`, as described in [the debugging documentation](https://docs.pennylane.ai/en/stable/code/qml_debugging.html#controlling-code-execution-in-the-debugging-context).
  
  Finally, to modify a circuit mid-run, simply call the desired PennyLane operations:

  ```pycon
  (pldb) qml.CNOT(wires=(0,2))
  CNOT(wires=[0, 2])
  (pldb): print(qml.debug_tape().draw(wire_order=[0,1,2]))
  0: ──H─╭●─╭●─┤  
  2: ────╰X─╰X─┤  
  ```

<h4>Convert between OpenFermion and PennyLane 🤝</h4>

* Two new functions called `qml.from_openfermion` and `qml.to_openfermion` are now available to convert between 
  OpenFermion and PennyLane objects. This includes both fermionic and qubit operators.
  [(#5773)](https://github.com/PennyLaneAI/pennylane/pull/5773)
  [(#5808)](https://github.com/PennyLaneAI/pennylane/pull/5808)
  [(#5881)](https://github.com/PennyLaneAI/pennylane/pull/5881)
  
  For fermionic operators:

  ```pycon

>   >>> of_fermionic = openfermion.FermionOperator('0^ 2')

Reviewer: ```suggestion
  >>> import openfermion
  >>> of_fermionic = openfermion.FermionOperator('0^ 2')
```

  calculations. Rather than providing an integer value to `level`, it is possible to target
  the `"user"`, `"gradient"` or `"device"` stages:

  ```python
  n_wires = 3
  x = np.random.random((2, n_wires))

  @qml.qnode(qml.device("default.qubit"))
  def f():
      qml.BasicEntanglerLayers(x, range(n_wires))
      return qml.expval(qml.X(0))
  ```

  ```pycon
  >>> print(qml.draw(f, level="device")())
  0: ──RX(0.28)─╭●────╭X──RX(0.70)─╭●────╭X─┤  <X>
  1: ──RX(0.52)─╰X─╭●─│───RX(0.65)─╰X─╭●─│──┤     
  2: ──RX(0.00)────╰X─╰●──RX(0.03)────╰X─╰●─┤     
  ```

<h3>Improvements 🛠</h3>

<h4>Community contributions, including UnitaryHACK 💛</h4>

* `default.clifford` now supports arbitrary state-based measurements with `qml.Snapshot`.
  [(#5794)](https://github.com/PennyLaneAI/pennylane/pull/5794)

* `qml.equal` now properly handles `Pow`, `Adjoint`, `Exp`, and `SProd` operators as arguments across 
  different interfaces and tolerances with the addition of four new keyword arguments: `check_interface`, 
  `check_trainability`, `atol` and `rtol`.
  [(#5668)](https://github.com/PennyLaneAI/pennylane/pull/5668)
  

> * The sorting order of parameter-shift terms is now guaranteed to resolve ties in the absolute value with the sign of the shifts.
>   [(#5583)](https://github.com/PennyLaneAI/pennylane/pull/5583)

Reviewer: ```suggestion
```
Not a community contrib

<h3>New features since last release</h3>

<h4>Execute wide circuits with Default Tensor 🔗</h4>

* A new `default.tensor` device is now available for performing
  [tensor network](https://en.wikipedia.org/wiki/Tensor_network) and
  [matrix product state](https://en.wikipedia.org/wiki/Matrix_product_state) simulations
  of quantum circuits using the
  [quimb backend](https://quimb.readthedocs.io/en/latest/).
  [(#5699)](https://github.com/PennyLaneAI/pennylane/pull/5699)
  [(#5744)](https://github.com/PennyLaneAI/pennylane/pull/5744)
  [(#5786)](https://github.com/PennyLaneAI/pennylane/pull/5786)
  [(#5795)](https://github.com/PennyLaneAI/pennylane/pull/5795)

>   Either method can be selected when instantiating the `default.tensor` device by setting the
>   `method` keyword argument to `"tn"` (tensor network) or `"mps"` (matrix product state). This
>   device can simulate a large number of qubits with sensible approximations. Take this example that
>   can simulate 1000 qubits in a few seconds!

Reviewer 2: ```suggestion
  Either method can be selected when instantiating the `default.tensor` device by setting the
  `method` keyword argument to `"tn"` (tensor network) or `"mps"` (matrix product state). This
  device can simulate a large number of qubits. Take this example that
  simulates 1000 qubits with a shallow circuit in a few seconds!
```

perhaps like this?
Claiming that it can _efficiently_ simulate a large number of qubits without saying any caveats could be misinterpreted I think

Author: :shipit:

<h4>Execute faster with Default Tensor 🔗</h4>

* A new `default.tensor` device is now available for performing
  [tensor network](https://en.wikipedia.org/wiki/Tensor_network) and
  [matrix product state](https://en.wikipedia.org/wiki/Matrix_product_state) simulations using the

>   [`quimb` backend](https://quimb.readthedocs.io/en/latest/).

Reviewer: ```suggestion
  [quimb backend](https://quimb.readthedocs.io/en/latest/).
```

  [tensor network](https://en.wikipedia.org/wiki/Tensor_network) and
  [matrix product state](https://en.wikipedia.org/wiki/Matrix_product_state) simulations
  of quantum circuits using the
  [quimb backend](https://quimb.readthedocs.io/en/latest/).
  [(#5699)](https://github.com/PennyLaneAI/pennylane/pull/5699)
  [(#5744)](https://github.com/PennyLaneAI/pennylane/pull/5744)
  [(#5786)](https://github.com/PennyLaneAI/pennylane/pull/5786)
  [(#5795)](https://github.com/PennyLaneAI/pennylane/pull/5795)

  Either method can be selected when instantiating the `default.tensor` device by setting the
  `method` keyword argument to `"tn"` (tensor network) or `"mps"` (matrix product state). This
  device can simulate a large number of qubits:

  ```python
  import pennylane as qml

  num_qubits = 1000
  dev = qml.device("default.tensor", method="mps")

  @qml.qnode(dev)
  def circuit():
      qml.Hadamard(0)
      for i in range(num_qubits - 1):
          qml.CNOT([i, i + 1])
      return qml.expval(qml.Z(num_qubits - 1))

  circuit()
  ```

  Additional configuration settings like the matrix product state bond dimension can be specified
  when instantiating `default.tensor`, see the

>   [device documentation](https://docs.pennylane.ai/en/stable/code/api/pennylane.devices.default_tensor.DefaultTensor.html)

Reviewer: ```suggestion
  [device's documentation](https://docs.pennylane.ai/en/stable/code/api/pennylane.devices.default_tensor.DefaultTensor.html)
```

  for more details.
  These settings have implications for the speed and accuracy of the simulation, as shown in the
  following example. First, set up the circuit:

  ```python
  import numpy as np

  n_layers = 10
  n_wires = 10

  initial_shape, weights_shape = qml.SimplifiedTwoDesign.shape(n_layers, n_wires)
  np.random.seed(1967)
  initial_layer_weights = np.random.random(initial_shape)
  weights = np.random.random(weights_shape)

  def f():
      qml.SimplifiedTwoDesign(initial_layer_weights, weights, range(n_wires))
      return qml.expval(qml.Z(0))
  ```

  The `default.tensor` device is instantiated with a `max_bond_dim` value:

  ```python

  dev_dq = qml.device("default.qubit")
  value_dq = qml.QNode(f, dev_dq)()

  dev_mps = qml.device("default.tensor", max_bond_dim=5)
  value_mps = qml.QNode(f, dev_mps)()
  ```

  With this bond dimension, the expectation values calculated for `default.qubit` and
  `default.tensor` are different:

  ```pycon
  >>> np.abs(value_dq - value_mps)
  tensor(0.0253213, requires_grad=True)
  ```

>   Learn more about `default.tensor` and how to configure it by visiting the
>   [how-to guide](https://pennylane.ai/qml/demos/tutorial_How_to_simulate_quantum_circuits_with_tensor_networks/).

Reviewer: ```suggestion
  Learn more about `default.tensor` and how to configure it by 
  [visiting the how-to guide](https://pennylane.ai/qml/demos/tutorial_How_to_simulate_quantum_circuits_with_tensor_networks/).
```

> <h4>Execute faster with Default Tensor 🔗</h4>

Reviewer 2: ```suggestion
<h4>Execute bigger with Default Tensor 🔗</h4>
```

I would suggest this reframing since the main point of the device is size and not speed. In fact, most <20q circuits are running slower on default.tensor than default.qubit, as the point is more scalability (at this point at least)

Author: Fair point! Wasn't aware of DT being slower than DQ for <20 qubits 🤔... Is this strictly implementation that's causing this? I would expect that I could at least get super fast results with a crappy bond dimension for upwards of ~12 qubits, no?

Reviewer 2: Yeah, the devil is in the detail: So for "typical" circuits you have to go to larger qubit sizes to get a chance to beat default.qubit, even for small bond dimensions. There are some known performance bottlenecks from re-canonicalizing the MPS state that we havent addressed yet, plus overall there are some competing overheads from SVDs that make it slower.

Check this example with 8 qubits. Even with trivial bond dimension of 2 DQ is faster:
```
dataset = qml.data.load("qchem", molname="H4", bondlength=0.92, basis="STO-3G")[0]

H = dataset.hamiltonian
n_wires = len(H.wires)

def circuit():
    qml.BasisState(dataset.hf_state, wires=H.wires)
    for op in dataset.vqe_gates:
        qml.apply(op)
    return qml.expval(H)

bond_dims = 2**np.arange(2, (n_wires//2)+1) # maximum required bond dinemsion is 2**(n_wires//2) = 64
ress = []

bond_dim = 2
mps = qml.device("default.tensor", wires=n_wires, method="mps", max_bond_dim=bond_dim, contract="nonlocal")
dev = qml.device("default.qubit")
qnode_mps = qml.QNode(circuit, mps)
qnode_def = qml.QNode(circuit, dev)

%timeit res = qnode_mps()
179 ms ± 5.34 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
%timeit res = qnode_def()
116 ms ± 3.49 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

This is one example of how performance can look like (here the circuit depth is constant)
![image](https://github.com/PennyLaneAI/pennylane/assets/43949391/89cea045-7963-439d-9c91-f9e61df2cb33)
The take away is that typically the break-even point is somewhere between 12-20 qubits, depending on the circuit and bond dimension. So people switching out the device for small examples will likely first see a performance regression (before they scale up qubit count)

Author: ```suggestion
<h4>Execute wide circuits with Default Tensor 🔗</h4>
```

* A new `default.tensor` device is now available for performing
  [tensor network](https://en.wikipedia.org/wiki/Tensor_network) and
  [matrix product state](https://en.wikipedia.org/wiki/Matrix_product_state) simulations
  of quantum circuits using the
  [quimb backend](https://quimb.readthedocs.io/en/latest/).
  [(#5699)](https://github.com/PennyLaneAI/pennylane/pull/5699)
  [(#5744)](https://github.com/PennyLaneAI/pennylane/pull/5744)
  [(#5786)](https://github.com/PennyLaneAI/pennylane/pull/5786)
  [(#5795)](https://github.com/PennyLaneAI/pennylane/pull/5795)

  Either method can be selected when instantiating the `default.tensor` device by setting the

>   `method` keyword argument to `"tn"` (tensor network) or `"mps"` (matrix product state). This
>   device can simulate a large number of qubits with sensible approximations. Take this example that
>   can simulate 1000 qubits in a few seconds!

Reviewer 2: ```suggestion
  `method` keyword argument to `"tn"` (tensor network) or `"mps"` (matrix product state). This
  device can simulate a large number of qubits with low-to-intermediate entanglement. Take this example that
  can simulate 1000 qubits with a shallow circuit in a few seconds!
```

The full contraction method (in its current form) is always exact and there is no approximation. Both methods have different strengths, but it would be correct to generally say for both that they are good for simulating low-to-intermediately entangled circuits. "Low-to-intermediate" is a bit of a mouthful an vague, could be replaced by something else in case you dont like it :)

Author: This is always a tough thing to describe without going into too much detail, so the nuance I left out with what's currently written is probably confusing you — totally understandable 😅. The point I'm making is akin to saying "trotterization is efficient with sensible approximations". The implementation of the method under the hood in PennyLane is exact, but certainly the trotterization technique is approximating something. 

So, with MPS, the implementation of the contraction method is exact, and anything under the sun can be represented by an MPS... _with a big enough bond dimension_, and that's what I'm talking about when I say "approximations". If a crappy bond dimension is chosen, the result is an approximation of the exact result. 

(_takes breath_)

What's probably better to say here is this (leave even more specifics out 😆, and let the user discover the wonderful world of matrix product states)

```suggestion
  `method` keyword argument to `"tn"` (tensor network) or `"mps"` (matrix product state). This
  device can efficiently simulate a large number of qubits. Take this example that
  can simulate 1000 qubits in a few seconds!
```

The reason I don't love "low-to-intermediate entanglement" here is because you can have lots of entanglement so long as we have area-law scaling. 

TLDR; let's just be super vague 😆

Reviewer 2: I get your point about the MPS method, just want to clarify that the full contraction method (`method="tn"`) really is exact-exact and there is (currently) no way to introduce approximations.

Author: Oooooooooh okay okay... apologies 😆 I over explained the crap out of my perspective then lol. In that case, I still prefer my suggestion above. You good with that? (emphasis on the "can" in "can efficiently simulate ...")

  ```python
  import pennylane as qml

  num_qubits = 1000
  dev = qml.device("default.tensor", method="mps")

  @qml.qnode(dev)
  def circuit():
      qml.Hadamard(0)
      for i in range(num_qubits - 1):
          qml.CNOT([i, i + 1])
      return qml.expval(qml.Z(num_qubits - 1))
  ```

  ```pycon
  >>> circuit()
  tensor(0., requires_grad=True)
  ```

>   Additional configuration settings like the matrix product state bond dimension can be specified
>   when instantiating `default.tensor` (see the
>   [device's documentation](https://docs.pennylane.ai/en/stable/code/api/pennylane.devices.default_tensor.DefaultTensor.html)
>   for more details).
>   These settings have implications for the speed and accuracy of the simulation, as shown in the
>   following example. First, set up the circuit:

Reviewer 2: ```suggestion
  For matrix product state simulations (`method="mps"`), we can make the exeuction approximate by setting `max_bond_dim` (see the
  [device's documentation](https://docs.pennylane.ai/en/stable/code/api/pennylane.devices.default_tensor.DefaultTensor.html)
  for more details).
  The maximum bond dimension lets us control the degree of the approximation and has implications for the speed of the simulation, as shown in the
  following example. First, set up the circuit:
```

Some details as the default for `max_bond_dim` is such that the execution is always exact.

  ```python
  import numpy as np

  n_layers = 10
  n_wires = 10

  initial_shape, weights_shape = qml.SimplifiedTwoDesign.shape(n_layers, n_wires)
  np.random.seed(1967)
  initial_layer_weights = np.random.random(initial_shape)
  weights = np.random.random(weights_shape)

  def f():
      qml.SimplifiedTwoDesign(initial_layer_weights, weights, range(n_wires))
      return qml.expval(qml.Z(0))
  ```

  The `default.tensor` device is instantiated with a `max_bond_dim` value:

  ```python
  dev_dq = qml.device("default.qubit")
  value_dq = qml.QNode(f, dev_dq)()

  dev_mps = qml.device("default.tensor", max_bond_dim=5)
  value_mps = qml.QNode(f, dev_mps)()
  ```

  With this bond dimension, the expectation values calculated for `default.qubit` and
  `default.tensor` are different:

  ```pycon
  >>> np.abs(value_dq - value_mps)
  tensor(0.0253213, requires_grad=True)

>   ```

Author: Should we add some timing results here as well?

* `default.clifford` now supports arbitrary state-based measurements with `qml.Snapshot`.
  [(#5794)](https://github.com/PennyLaneAI/pennylane/pull/5794)

> * `qml.equal` now properly handles `Pow`, `Adjoint`, `Exp`, and `SProd` operators as arugments across 

Reviewer: ```suggestion
* `qml.equal` now properly handles `Pow`, `Adjoint`, `Exp`, and `SProd` operators as arguments across 
```

  the end of the circuit, just like `qml.from_qiskit`.
  [(#5818)](https://github.com/PennyLaneAI/pennylane/pull/5818)

* Four new operators have been added for simulating noise on the `default.qutrit.mixed` device:
  [(#5502)](https://github.com/PennyLaneAI/pennylane/pull/5502)
  [(#5793)](https://github.com/PennyLaneAI/pennylane/issues/5793)
  [(#5503)](https://github.com/PennyLaneAI/pennylane/pull/5503)
  [(#5757)](https://github.com/PennyLaneAI/pennylane/pull/5757)
  [(#5799)](https://github.com/PennyLaneAI/pennylane/pull/5799)
  [(#5784)](https://github.com/PennyLaneAI/pennylane/pull/5784)

  * `qml.QutritDepolarizingChannel`: a channel that adds depolarizing noise.
  * `qml.QutritChannel`: enables the specification of noise using a collection of (3x3) Kraus matrices.
  * `qml.QutritAmplitudeDamping`: a channel that adds noise processes modelled by amplitude damping.
  * `qml.TritFlip`: a channel that adds trit flip errors, such as misclassification.

<h4>Faster and more flexible mid-circuit measurements</h4>

* The `default.qubit` device supports a depth-first tree-traversal algorithm to accelerate native mid-circuit 

>   measurement execution. The new implementation supports classical control, collecting statistics, and 

Reviewer: ```suggestion
  measurement execution. Accessible through the QNode argument `mcm_method="tree-traversal"`,
  this new implementation supports classical control, collecting statistics, and 
```

  post-selection, along with all measurements enabled with `qml.dynamic_one_shot`. More information 
  about this new mid-circuit measurement method can be found on our 

>   [measurement documentation page](https://docs.pennylane.ai/en/stable/introduction/measurements.html).

Reviewer: ```suggestion
  [measurement documentation page](https://docs.pennylane.ai/en/stable/introduction/dynamic_quantum_circuits.html#tree-traversal-algorithm).
```

    to execute once for each shot. If `qml.qjit` is being used (the Catalyst compiler), `mcm_method="single-branch-statistics"`
    is also available. Using this method, a single branch of the execution tree will be randomly explored.

* The `dynamic_one_shot` transform received a few improvements:
  
  * `dynamic_one_shot` is now compatible with `qml.qjit` (the Catalyst compiler).
    [(#5766)](https://github.com/PennyLaneAI/pennylane/pull/5766)
    [(#5888)](https://github.com/PennyLaneAI/pennylane/pull/5888)
  * `dynamic_one_shot` now uses a single auxiliary tape with a shot vector and `default.qubit` implements 
    the loop over shots with `jax.vmap`.
    [(#5617)](https://github.com/PennyLaneAI/pennylane/pull/5617)
  * `dynamic_one_shot` is now compatible with `jax.jit`.
    [(#5557)](https://github.com/PennyLaneAI/pennylane/pull/5557)

* Mid-circuit measurement tests have been streamlined and refactored, removing most end-to-end tests 
  from the native MCM test file, but keeping one that validates multiple mid-circuit measurements with 
  any allowed return and interface end-to-end tests.

>   [(#5787)](https://github.com/PennyLaneAI/pennylane/pull/5787)

Reviewer: Minor, but this feels not to exciting and maybe worth moving to the bottom of this section.

      return qml.sample(wires = [2,3,4])
  ```
  
  ```pycon
  >>> print(circuit())
  [1 1 0]
  ```

<h4>Capturing and representing hybrid programs</h4>

* A number of templates have been updated to be valid PyTrees and PennyLane operations.
  [(#5698)](https://github.com/PennyLaneAI/pennylane/pull/5698)

* PennyLane operators, measurements, and QNodes can now automatically be captured as instructions in 
  JAXPR.
  [(#5564)](https://github.com/PennyLaneAI/pennylane/pull/5564)
  [(#5511)](https://github.com/PennyLaneAI/pennylane/pull/5511)
  [(#5708)](https://github.com/PennyLaneAI/pennylane/pull/5708)
  [(#5523)](https://github.com/PennyLaneAI/pennylane/pull/5523)
  [(#5686)](https://github.com/PennyLaneAI/pennylane/pull/5686)
  [(#5889)](https://github.com/PennyLaneAI/pennylane/pull/5889)

* The `qml.PyTrees` module now has `flatten` and `unflatten` methods for serializing PyTrees.
  [(#5701)](https://github.com/PennyLaneAI/pennylane/pull/5701)

> * `qml.sample` can now be used on boolean values representing mid-circuit measurement results in

Reviewer: I think it is normally capitalized

Author: ```suggestion
* `qml.sample` can now be used on Boolean values representing mid-circuit measurement results in
```

  `"dhf"` and `"openfermion"` backends.
  [(#5764)](https://github.com/PennyLaneAI/pennylane/pull/5764)

* The qchem module now has dedicated functions for calling the `pyscf` and `openfermion` backends and 
  the `molecular_hamiltonian` and `molecular_dipole` functions have been moved to `hamiltonian` and
  `dipole` modules.
  [(#5553)](https://github.com/PennyLaneAI/pennylane/pull/5553)
  [(#5863)](https://github.com/PennyLaneAI/pennylane/pull/5863)

* More fermionic-to-qubit tests have been added to cover cases when the mapped operator is different 
  for various mapping schemes.
  [(#5873)](https://github.com/PennyLaneAI/pennylane/pull/5873)

<h4>Easier development</h4>

* Logging now allows for an easier opt-in across the stack and support has been extended to Catalyst.
  [(#5528)](https://github.com/PennyLaneAI/pennylane/pull/5528)

* Three new Pytest markers have been added for easier management of our test suite: `unit`, `integration` 
  and `system`.
  [(#5517)](https://github.com/PennyLaneAI/pennylane/pull/5517)

<h4>Other improvements</h4>

* `qml.MultiControlledX` can now be decomposed even when no `work_wires` are provided. The implementation 

>   returns $\mathcal{O}(\text{len(control\_wires)}^2)$ operations and is applicable for any multi-controlled 

Reviewer: ```suggestion
  returns :math:`\mathcal{O}(\text{len(control\_wires)}^2)` operations and is applicable for any multi-controlled 
```
Doesn't currently show up correctly in the [docs](https://xanaduai-pennylane--5928.com.readthedocs.build/en/5928/development/release_notes.html)

  `"dhf"` and `"openfermion"` backends.
  [(#5764)](https://github.com/PennyLaneAI/pennylane/pull/5764)

* The qchem module now has dedicated functions for calling the `pyscf` and `openfermion` backends and 
  the `molecular_hamiltonian` and `molecular_dipole` functions have been moved to `hamiltonian` and
  `dipole` modules.
  [(#5553)](https://github.com/PennyLaneAI/pennylane/pull/5553)
  [(#5863)](https://github.com/PennyLaneAI/pennylane/pull/5863)

* More fermionic-to-qubit tests have been added to cover cases when the mapped operator is different 
  for various mapping schemes.
  [(#5873)](https://github.com/PennyLaneAI/pennylane/pull/5873)

<h4>Easier development</h4>

* Logging now allows for an easier opt-in across the stack and support has been extended to Catalyst.
  [(#5528)](https://github.com/PennyLaneAI/pennylane/pull/5528)

* Three new Pytest markers have been added for easier management of our test suite: `unit`, `integration` 
  and `system`.
  [(#5517)](https://github.com/PennyLaneAI/pennylane/pull/5517)

<h4>Other improvements</h4>

* `qml.MultiControlledX` can now be decomposed even when no `work_wires` are provided. The implementation 

>   returns :math:`\mathcal{O}(\mbox{len(control\_wires)}^2)` operations and is applicable for any multi-controlled 

Reviewer: ```suggestion
  returns :math:`\mathcal{O}(\mbox{len(control wires)}^2)` operations and is applicable for any multi-controlled 
```

  `"dhf"` and `"openfermion"` backends.
  [(#5764)](https://github.com/PennyLaneAI/pennylane/pull/5764)

* The qchem module now has dedicated functions for calling the `pyscf` and `openfermion` backends and 
  the `molecular_hamiltonian` and `molecular_dipole` functions have been moved to `hamiltonian` and
  `dipole` modules.
  [(#5553)](https://github.com/PennyLaneAI/pennylane/pull/5553)
  [(#5863)](https://github.com/PennyLaneAI/pennylane/pull/5863)

* More fermionic-to-qubit tests have been added to cover cases when the mapped operator is different 
  for various mapping schemes.
  [(#5873)](https://github.com/PennyLaneAI/pennylane/pull/5873)

<h4>Easier development</h4>

* Logging now allows for an easier opt-in across the stack and support has been extended to Catalyst.
  [(#5528)](https://github.com/PennyLaneAI/pennylane/pull/5528)

* Three new Pytest markers have been added for easier management of our test suite: `unit`, `integration` 
  and `system`.
  [(#5517)](https://github.com/PennyLaneAI/pennylane/pull/5517)

<h4>Other improvements</h4>

* `qml.MultiControlledX` can now be decomposed even when no `work_wires` are provided. The implementation 

>   returns :math:`\mathcal{O}(\text{len(control\_wires)}^2)` operations and is applicable for any multi-controlled 

Reviewer: ```suggestion
  returns :math:`\mathcal{O}(\mbox{len(control\_wires)}^2)` operations and is applicable for any multi-controlled 
```

* `qml.TrotterProduct` is now compatible with resource tracking by inheriting from `ResourcesOperation`. 

>    [(#5680)](https://github.com/PennyLaneAI/pennylane/pull/5680)

Reviewer: ```suggestion
  [(#5680)](https://github.com/PennyLaneAI/pennylane/pull/5680)
```

  functionalities. Other properties such as adjoint and diagonalizing gates have been added using the 
  `base` properties.
  [(##5772)](https://github.com/PennyLaneAI/pennylane/pull/5772)

* New dispatches for `qml.ops.Conditional` and `qml.MeasurementValue` have been added to `qml.equal`.
  [(##5772)](https://github.com/PennyLaneAI/pennylane/pull/5772)

* The `qml.snapshots` transform now supports arbitrary devices by running a separate tape for each snapshot 
  for unsupported devices.
  [(#5805)](https://github.com/PennyLaneAI/pennylane/pull/5805)

* The `qml.Snapshot` operator now accepts sample-based measurements for finite-shot devices.
  [(#5805)](https://github.com/PennyLaneAI/pennylane/pull/5805)

* Device preprocess transforms now happen inside the ML boundary.
  [(#5791)](https://github.com/PennyLaneAI/pennylane/pull/5791)

* Transforms applied to callables now use `functools.wraps` to preserve the docstring and call signature 
  of the original function.
  [(#5857)](https://github.com/PennyLaneAI/pennylane/pull/5857)

* `qml.qsvt()` now supports JAX arrays with angle conversions. 
  [(#5853)](https://github.com/PennyLaneAI/pennylane/pull/5853)

> 

Reviewer: ```suggestion

* The sorting order of parameter-shift terms is now guaranteed to resolve ties in the absolute value with the sign of the shifts.
  [(#5583)](https://github.com/PennyLaneAI/pennylane/pull/5583)

```
